File: Python/020_Valid_Parentheses.py
class Solution(object):
    def isValid(self, s):
        if len(s) % 2 != 0 :
            return False
        mapp = {'(' : 1 , '{' : 2 , '[' : 3 , ']' : 4 , '}' : 5 , ')' : 6 }
        stack = []
        for item in s :
            if mapp[item] <= 3 :
                stack.append(item)
            else :
                temp = mapp[item]
                tmp = 0
                while True :
                    if len(stack) == 0 :
                        if tmp % 7 != 0 :
                            return False
                        break
                    k = stack.pop()
                    if mapp[k] + temp == 7 :
                        if tmp % 7 != 0 :
                            return False
                        break
                    tmp += mapp[k]
        if len(stack) != 0 :
            return False
        return True

File: C++/020_Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if ( s.length() % 2 ) 
            return false ; 
        stack<int> z ; 
        map<char,int> mmp ; 
        string temp = "({[]})" ; 
        for ( int i = 0 ; i < 6 ; i++ ) { 
            mmp[temp[i]] = i + 1  ; 
        }
        for ( int i = 0 ; i < s.length() ; i ++ ) {
            if ( mmp[s[i]] <= 3 ) 
                z.push(mmp[s[i]]) ; 
            else  { 
                int sum = 0 ; 
                while ( z.size() ) { 
                    int tmp = z.top() ; 
                    z.pop() ; 
                    if ( tmp + mmp[s[i]] == 7 ) {
                        if ( sum % 7 ) 
                            return false ; 
                        break ; 
                    } 
                    if ( z.size() == 0 ) 
                        return false ; 
                    sum += tmp ;
                }
            }
        }
        if ( z.size() != 0 )
            return false ; 
        return true ; 
    }
};